tensorflow/contrib/avro/python/avro_record_dataset.py

```diff
@@ -0,0 +1,222 @@
+    #   input_arg {
+    #     name: "filenames"
+    #     type: DT_STRING
+    #   }
+    #   input_arg {
+    #     name: "schema"
+    #     type: DT_STRING
+    #   }
+    #   input_arg {
+    #     name: "buffer_size"
+    #     type: DT_INT64
+    #   }
+    #   output_arg {
+    #     name: "handle"
+    #     type: DT_VARIANT
+    #   }
+    #   is_stateful: true
+    # }
+
+    # Create the operator definition
+    avro_dataset_op = _op_def_pb2.OpDef()
+    avro_dataset_op.name = "AvroRecordDataset"
+
+    # Create and add the input arguments
+    input_filenames = _op_def_pb2.OpDef.ArgDef()
+    input_filenames.name = "filenames"
+    input_filenames.type = _types_pb2.DT_STRING
+    input_schema = _op_def_pb2.OpDef.ArgDef()
+    input_schema.name = "schema"
+    input_schema.type = _types_pb2.DT_STRING
+    input_buffer_size = _op_def_pb2.OpDef.ArgDef()
+    input_buffer_size.name = "buffer_size"
+    input_buffer_size.type = _types_pb2.DT_INT64
+    avro_dataset_op.input_arg.extend([input_filenames, input_schema, input_buffer_size])
+
+    # Create and add the output argument
+    output_handle = _op_def_pb2.OpDef.ArgDef()
+    output_handle.name = "handle"
+    output_handle.type = _types_pb2.DT_VARIANT
+    avro_dataset_op.output_arg.extend([output_handle])
+
+    # This method is stateful because it keeps a counter into rows/records
+    avro_dataset_op.is_stateful = True
+
+    return avro_dataset_op
+
+
+def _create_op_def_library(op_proto):
+    """
+    Creates a customized operator definition library for the given op_proto.
+
+    Notice that this method will check that a CC implementation for this method with the SAME signature has been
+    registered with TensorFlow.  Make sure to load the native code before calling this method.
+
+    :return: A operation definition library.
+    """
+    # Log the proto representation of the operator
+    logging.info("Adding operator {} to the library.".format(op_proto))
+
+    # Register the new op with TF and create an op library for it
+    # Note: The native cc code will register the operator in _registered_ops, This is for checking
+    registered_ops = _registry.get_registered_ops()
+    if op_proto.name not in registered_ops:
+        raise RuntimeError("Could not find native implementation of op '{}'.".format(op_proto.name))
+
+    op_def_lib = _op_def_library.OpDefLibrary()
+    ops_proto = _op_def_pb2.OpList()
+    ops_proto.op.extend([op_proto])
+
+    # Make sure the registered op matches this one by re-registering it
+    _registry.register_op_list(ops_proto)  # Will fail if the natively registered op differs
+
+    # Add the op to the customized operator library
+    op_def_lib.add_op_list(ops_proto)
+
+    return op_def_lib
+
+
+# Load the customized operator library which contains only our avro dataset operator -- but that is enough here
+_op_def_lib = _create_op_def_library(_create_avro_dataset_op_proto())
+
+
+def _convert_optional_param_to_tensor(argument_name,
```

Comment:
Why not import tensorflow.python.data.util.convert instead of copying?

Reply:
Changed.